services/scabbard/libscabbard/src/store/scabbard_store/diesel/operations/add_consensus_action.rs

```diff
@@ -0,0 +1,668 @@
+        consensus_update_coordinator_context_action,
+        consensus_update_coordinator_context_action_participant,
+        consensus_update_participant_context_action,
+        consensus_update_participant_context_action_participant,
+    },
+};
+use crate::store::scabbard_store::ScabbardStoreError;
+use crate::store::scabbard_store::{
+    two_phase::{
+        action::{ConsensusAction, ConsensusActionNotification},
+        message::Scabbard2pcMessage,
+    },
+    ScabbardConsensusAction, ScabbardContext,
+};
+
+use super::ScabbardStoreOperations;
+
+pub(in crate::store::scabbard_store::diesel) trait AddActionOperation {
+    fn add_consensus_action(
+        &self,
+        action: ScabbardConsensusAction,
+        service_id: &FullyQualifiedServiceId,
+        epoch: u64,
+    ) -> Result<i64, ScabbardStoreError>;
+}
+
+#[cfg(feature = "sqlite")]
+impl<'a> AddActionOperation for ScabbardStoreOperations<'a, SqliteConnection> {
+    fn add_consensus_action(
+        &self,
+        action: ScabbardConsensusAction,
+        service_id: &FullyQualifiedServiceId,
+        epoch: u64,
+    ) -> Result<i64, ScabbardStoreError> {
+        self.conn.transaction::<_, _, _>(|| {
+            let ScabbardConsensusAction::Scabbard2pcConsensusAction(action) = action;
+            let epoch = i64::try_from(epoch).map_err(|err| {
+                ScabbardStoreError::Internal(InternalError::from_source(Box::new(err)))
+            })?;
+            // check to see if a coordinator context with the given epoch and service_id exists
+            let coordinator_context =
+                consensus_coordinator_context::table
+                    .filter(consensus_coordinator_context::epoch.eq(epoch).and(
+                        consensus_coordinator_context::service_id.eq(format!("{}", service_id)),
+                    ))
+                    .first::<CoordinatorContextModel>(self.conn)
+                    .optional()?;
+
+            // check to see if a participant context with the given epoch and service_id exists
+            let participant_context =
+                consensus_participant_context::table
+                    .filter(consensus_participant_context::epoch.eq(epoch).and(
+                        consensus_participant_context::service_id.eq(format!("{}", service_id)),
+                    ))
+                    .first::<ParticipantContextModel>(self.conn)
+                    .optional()?;
+
+            let position = consensus_action::table
+                .filter(
+                    consensus_action::service_id
+                        .eq(format!("{}", service_id))
+                        .and(consensus_action::epoch.eq(epoch)),
+                )
+                .order(consensus_action::position.desc())
+                .select(consensus_action::position)
+                .first::<i32>(self.conn)
+                .optional()?
+                .unwrap_or(0)
+                + 1;
```

Comment:
Does this plus +1 work if multiple actions are being added atomically at the same time?

Reply:
wrote a test for this that creates 200 actions using store commands and adds them to the db atomically using the executor and they were all assigned position correctly, haven't pushed up the test yet though because of the issue I mentioned when trying to pull in the necessary splinter features